Implement pen blocks

cwillisf opened the pull request:
### Resolves

Along with LLK/scratch-render#72, this resolves LLK/scratch-render#6

### Proposed Changes

These blocks implement pen features as found in Scratch 2.0

Supporting changes include:
- `Target` is now an event emitter
- `RenderedTarget` now emits an event when it moves
- `Target` can now store additional arbitrary data, called "custom state" in the code, using a `Target`'s `setCustomState` and `getCustomState` methods. This is used to store per-target pen state without requiring `Target` or `RenderedTarget` to know anything about the pen.
- `Cast` can now cast to an RGB color object.
- `Color` now has functions to convert between RGB and HSV, constants for for black & white, and a `mixRgb` function to lerp between two colors.

### Reason for Changes

Implementing pen blocks brings us closer to Scratch 2.0 compatibility.

### Test Coverage

All newly-added `util` functions are covered:
- `toRbgColorObject` from `util/cast.js` is covered in `test/unit/util_cast.js`
- `hsvToRgb`, `rgbToHsv`, and `mixRgb` from `util/color.js` are covered in `test/unit/util_color.js`

All pen blocks are covered by the new `pen.js` integration test. Unfortunately our current integration test setup makes it very difficult (impossible?) to hook to the renderer, so the blocks which actually draw something will skip parts of their functionality during this test. The three affected pen blocks are `clear`, `pen down`, and `stamp`.

Commit:
Implement pen blocks

These blocks implement pen features as found in Scratch 2.0

Supporting changes include:
- `Target` is now an event emitter
- `RenderedTarget` now emits an event when it moves
- `Target` can now store arbitrary "extra" data, called "custom state"
  in the code, using a `Target`'s `setCustomState` and `getCustomState`
  methods. This is used to store per-target pen state without requiring
  `Target` or `RenderedTarget` to know anything about the pen.
- `Cast` can now cast to an RGB color object.
- `Color` now has functions to convert between RGB and HSV, constants
  for for black & white, and a `mixRgb` function to lerp between two
  colors.

Commit:
Lint fixes and related cleanup

towerofnix commented on src/util/color.js:
```diff
@@ -99,28 +99,54 @@ Color.hexToDecimal = function (hex) {
+        g = p;
+        b = q;
+        break;
+    }
```
Hmph. This is (arguably) less readable than before, part because everything's on separate lines, somewhat away from the `case`s, and part because it's a switch-case, which gets you all the cludge of `case` and `break`...

Nitpicks :)

cwillisf replied:
Yeah... I'm open to ideas if you can think of a way to make it more readable while following the linting rules :)
I have some ideas but I'm not sure they'd work out. I probably won't mess with it right away.

Commit:
Add unit tests for new color-related functionality

Also, fix a math error in `Color.rgbToHsv`.

Newly covered functions:
- `toRbgColorObject` from `util/cast.js`
- `hsvToRgb` from `util/colors.js`
- `rgbToHsv` from `util/colors.js`
- `mixRgb` from `util/colors.js`

Commit:
Add pen integration test

See `pen.sb2` for details

thisandagain commented on src/blocks/scratch3_pen.js:
```diff
@@ -0,0 +1,231 @@
+// Place the pen layer in front of the backdrop but behind everything else
+// We should probably handle this somewhere else... somewhere central that knows about pen, backdrop, video, etc.
+// Maybe it should be in the GUI?
+var penOrder = 1;
```
We have generally been handling things like this and the `stateKey` (below) like this:
https://github.com/LLK/scratch-vm/blob/develop/src/engine/sequencer.js#L23

thisandagain commented on src/blocks/scratch3_pen.js:
```diff
@@ -0,0 +1,231 @@
+     */
+    this.runtime = runtime;
+
+    this._penSkinId = -1;
```
Missing jsdoc.

thisandagain commented on src/blocks/scratch3_pen.js:
```diff
@@ -0,0 +1,231 @@
+ * @private
+ */
+Scratch3PenBlocks.prototype._clampPenSize = function (requestedSize) {
+    return MathUtil.clamp(requestedSize, 1, 255);
```
Should we declare the max pen size at the top of the file?

thisandagain commented on src/blocks/scratch3_pen.js:
```diff
@@ -0,0 +1,231 @@
+    return MathUtil.clamp(requestedSize, 1, 255);
+};
+
+Scratch3PenBlocks.prototype._getPenLayerID = function () {
```
Missing jsdoc.

thisandagain commented on src/blocks/scratch3_pen.js:
```diff
@@ -0,0 +1,231 @@
+Scratch3PenBlocks.prototype._getPenState = function (target) {
+    var penState = target.getCustomState(stateKey);
+    if (!penState) {
+        penState = JSON.parse(JSON.stringify(defaultPenState));
```
Maybe we should create a "clone" util to make this easier to read?

thisandagain commented on src/blocks/scratch3_pen.js:
```diff
@@ -0,0 +1,231 @@
+
+Scratch3PenBlocks.prototype.clear = function () {
+    var penSkinId = this._getPenLayerID();
+    this.runtime.renderer.penClear(penSkinId);
```
I think we should always check to make sure the renderer exists. E.g.:
https://github.com/LLK/scratch-vm/blob/develop/src/sprites/rendered-target.js#L56

thisandagain commented on src/engine/target.js:
```diff
@@ -112,10 +129,20 @@ Target.prototype.lookupOrCreateList = function (name) {
  */
 Target.prototype.postSpriteInfo = function () {};
 
+Target.prototype.getCustomState = function (stateId) {
```
Missing jsdoc.

thisandagain commented on src/engine/target.js:
```diff
@@ -112,10 +129,20 @@ Target.prototype.lookupOrCreateList = function (name) {
+    return this._customState[stateId];
+};
+
+Target.prototype.setCustomState = function (stateId, newValue) {
```
Missing jsdoc.

thisandagain reviewed (changes requested):
This looks great overall. Thanks for expanding the test coverage. Some fairly minor notes for the most part.

Commit:
Code review: more docs, move constants, clone util

Changes include:
- Added missing JSDoc for items in `scratch3_pen.js` and `target.js`.
- Moved constants used by `Scratch3PenBlocks` into the class.
- Created a constant for minimum and maximum pen size.
- Added `util/clone.js` to host cloning functionality.
- Pen blocks now check for the renderer before trying to use it.
- The pen integration test covers all blocks, though `clear`, `stamp`,
  and `pen down` will skip some of their functionality when there is no
  renderer.

Commit:
Undo accidental harmless but unnecessary change

SillyInventor commented:
For pen clear to work you need to set the skin to dirty:
this.runtime.renderer._allSkins[penSkinId]._canvasDirty=true;
